File: packages/typescript/src/core/SocialSDK.ts
/**
 * SocialSDK — top-level factory that wires up the SessionManager and exposes
 * lazy-instantiated platform adapters plus a `generic()` factory for arbitrary
 * websites.
 *
 * Usage:
 *   const sdk = await SocialSDK.create({ browserbase, llm });
 *   await sdk.platforms.instagram.login();
 *   await sdk.platforms.instagram.post({ text: "...", media: ["./img.jpg"] });
 *
 *   const reddit = sdk.generic({ siteId: "reddit", baseUrl: "https://reddit.com" });
 *   await reddit.login({ loginUrl: "/login" });
 */

import type {
  PlatformName,
  SocialSDKConfig,
} from "./types.js";
import { ConfigError } from "./errors.js";
import { SessionManager } from "./session.js";
import { BasePlatform } from "./BasePlatform.js";
import { GenericSiteAdapter, type GenericSiteOpts } from "../generic/GenericSiteAdapter.js";

// Platform adapters — lazy-imported to avoid pulling all 7 into every bundle.
type AdapterCtor = new (
  ...args: ConstructorParameters<typeof BasePlatform>
) => BasePlatform;

interface AdapterRegistry {
  instagram: AdapterCtor;
  tiktok: AdapterCtor;
  x: AdapterCtor;
  facebook: AdapterCtor;
  youtube: AdapterCtor;
  linkedin: AdapterCtor;
  threads: AdapterCtor;
}

async function loadAdapters(): Promise<AdapterRegistry> {
  const [
    { InstagramAdapter },
    { TikTokAdapter },
    { XAdapter },
    { FacebookAdapter },
    { YouTubeAdapter },
    { LinkedInAdapter },
    { ThreadsAdapter },
  ] = await Promise.all([
    import("../platforms/instagram.js"),
    import("../platforms/tiktok.js"),
    import("../platforms/x.js"),
    import("../platforms/facebook.js"),
    import("../platforms/youtube.js"),
    import("../platforms/linkedin.js"),
    import("../platforms/threads.js"),
  ]);
  return {
    instagram: InstagramAdapter as unknown as AdapterCtor,
    tiktok: TikTokAdapter as unknown as AdapterCtor,
    x: XAdapter as unknown as AdapterCtor,
    facebook: FacebookAdapter as unknown as AdapterCtor,
    youtube: YouTubeAdapter as unknown as AdapterCtor,
    linkedin: LinkedInAdapter as unknown as AdapterCtor,
    threads: ThreadsAdapter as unknown as AdapterCtor,
  };
}

export interface PlatformAccess {
  instagram: BasePlatform;
  tiktok: BasePlatform;
  x: BasePlatform;
  facebook: BasePlatform;
  youtube: BasePlatform;
  linkedin: BasePlatform;
  threads: BasePlatform;
}

export class SocialSDK {
  private readonly config: SocialSDKConfig;
  private readonly sessionManager: SessionManager;
  private readonly adapterRegistry: AdapterRegistry;
  private readonly adapterCache = new Map<string, BasePlatform>();
  private readonly genericCache = new Map<string, GenericSiteAdapter>();

  /** Default account label when caller doesn't pass one. */
  public defaultAccount = "default";

  /** Typed accessor for the 7 specialized platforms. */
  public readonly platforms: PlatformAccess;

  private constructor(
    config: SocialSDKConfig,
    sessionManager: SessionManager,
    adapterRegistry: AdapterRegistry,
  ) {
    this.config = config;
    this.sessionManager = sessionManager;
    this.adapterRegistry = adapterRegistry;

    const sdk = this;
    const make = (p: PlatformName): BasePlatform => sdk.platform(p);

    // Lazy-initialized via getters so unused platforms never instantiate Stagehand.
    this.platforms = {
      get instagram() { return make("instagram"); },
      get tiktok() { return make("tiktok"); },
      get x() { return make("x"); },
      get facebook() { return make("facebook"); },
      get youtube() { return make("youtube"); },
      get linkedin() { return make("linkedin"); },
      get threads() { return make("threads"); },
    };
  }

  static async create(config: SocialSDKConfig): Promise<SocialSDK> {
    if (!config.browserbase?.apiKey || !config.browserbase?.projectId) {
      throw new ConfigError(
        "browserbase.apiKey and browserbase.projectId are required",
      );
    }
    if (!config.llm?.apiKey) {
      throw new ConfigError("llm.apiKey is required");
    }

    const sessionManager = new SessionManager({
      apiKey: config.browserbase.apiKey,
      projectId: config.browserbase.projectId,
      sessionDir: config.sessionDir,
    });
    const adapters = await loadAdapters();
    return new SocialSDK(config, sessionManager, adapters);
  }

  /** Get (or build) a platform adapter for a given account. */
  platform(name: PlatformName, account?: string): BasePlatform {
    const acct = account ?? this.defaultAccount;
    const key = `${name}::${acct}`;
    const cached = this.adapterCache.get(key);
    if (cached) return cached;

    const Ctor = this.adapterRegistry[name];
    if (!Ctor) {
      throw new ConfigError(`Unknown platform: ${name}`);
    }
    const adapter = new Ctor({
      platform: name,
      account: acct,
      sessionManager: this.sessionManager,
      onMfaChallenge: this.config.onMfaChallenge,
      llm: this.config.llm,
      headless: this.config.headless,
      logLevel: this.config.logLevel,
    });
    this.adapterCache.set(key, adapter);
    return adapter;
  }

  /** Build a generic-site adapter for any URL. */
  generic(opts: {
    siteId: string;
    baseUrl: string;
    account?: string;
  }): GenericSiteAdapter {
    const acct = opts.account ?? this.defaultAccount;
    const key = `generic:${opts.siteId}::${acct}`;
    const cached = this.genericCache.get(key);
    if (cached) return cached;

    const generic = new GenericSiteAdapter({
      siteId: opts.siteId,
      baseUrl: opts.baseUrl,
      account: acct,
      sessionManager: this.sessionManager,
      onMfaChallenge: this.config.onMfaChallenge,
      llm: this.config.llm,
      headless: this.config.headless,
      logLevel: this.config.logLevel,
    } as GenericSiteOpts);
    this.genericCache.set(key, generic);
    return generic;
  }

  /** Tear down all open Stagehand sessions. Browserbase contexts persist. */
  async close(): Promise<void> {
    const all: Promise<unknown>[] = [];
    for (const adapter of this.adapterCache.values()) {
      all.push(adapter.close().catch(() => undefined));
    }
    for (const generic of this.genericCache.values()) {
      all.push(generic.close().catch(() => undefined));
    }
    await Promise.all(all);
    this.adapterCache.clear();
    this.genericCache.clear();
  }
}

File: packages/python/stagehand_social/core/social_sdk.py
"""SocialSDK factory — mirror of packages/typescript/src/core/SocialSDK.ts."""

from __future__ import annotations

from typing import Optional

from stagehand_social.core.base_platform import BasePlatform
from stagehand_social.core.errors import ConfigError
from stagehand_social.core.session import SessionManager
from stagehand_social.core.types import PlatformName, SocialSDKConfig
from stagehand_social.generic.generic_site import GenericSiteAdapter


class SocialSDK:
    """Top-level factory for platform adapters and the generic-site adapter."""

    default_account: str = "default"

    def __init__(
        self,
        config: SocialSDKConfig,
        session_manager: SessionManager,
    ):
        self._config = config
        self._session_manager = session_manager
        self._adapter_cache: dict[str, BasePlatform] = {}
        self._generic_cache: dict[str, GenericSiteAdapter] = {}

    @classmethod
    async def create(cls, config: SocialSDKConfig) -> "SocialSDK":
        if not config.browserbase.api_key or not config.browserbase.project_id:
            raise ConfigError("browserbase.api_key and browserbase.project_id are required")
        if not config.llm.api_key:
            raise ConfigError("llm.api_key is required")

        session_manager = SessionManager(
            api_key=config.browserbase.api_key,
            project_id=config.browserbase.project_id,
            session_dir=config.session_dir,
        )
        return cls(config, session_manager)

    @property
    def platforms(self) -> "_PlatformAccess":
        return _PlatformAccess(self)

    def platform(self, name: PlatformName, account: Optional[str] = None) -> BasePlatform:
        acct = account or self.default_account
        key = f"{name}::{acct}"
        if key in self._adapter_cache:
            return self._adapter_cache[key]

        # Lazy import to avoid loading all adapters every time.
        if name == "instagram":
            from stagehand_social.platforms.instagram import InstagramAdapter as Ctor
        elif name == "tiktok":
            from stagehand_social.platforms.tiktok import TikTokAdapter as Ctor
        elif name == "x":
            from stagehand_social.platforms.x import XAdapter as Ctor
        elif name == "facebook":
            from stagehand_social.platforms.facebook import FacebookAdapter as Ctor
        elif name == "youtube":
            from stagehand_social.platforms.youtube import YouTubeAdapter as Ctor
        elif name == "linkedin":
            from stagehand_social.platforms.linkedin import LinkedInAdapter as Ctor
        elif name == "threads":
            from stagehand_social.platforms.threads import ThreadsAdapter as Ctor
        else:
            raise ConfigError(f"Unknown platform: {name}")

        adapter = Ctor(
            platform=name,
            account=acct,
            session_manager=self._session_manager,
            on_mfa_challenge=self._config.on_mfa_challenge,
            llm=self._config.llm,
            headless=self._config.headless,
            log_level=self._config.log_level,
        )
        self._adapter_cache[key] = adapter
        return adapter

    def generic(
        self,
        site_id: str,
        base_url: str,
        account: Optional[str] = None,
    ) -> GenericSiteAdapter:
        acct = account or self.default_account
        key = f"generic:{site_id}::{acct}"
        if key in self._generic_cache:
            return self._generic_cache[key]
        generic = GenericSiteAdapter(
            site_id=site_id,
            base_url=base_url,
            account=acct,
            session_manager=self._session_manager,
            on_mfa_challenge=self._config.on_mfa_challenge,
            llm=self._config.llm,
            headless=self._config.headless,
            log_level=self._config.log_level,
        )
        self._generic_cache[key] = generic
        return generic

    async def close(self) -> None:
        import asyncio

        coros: list = []
        for adapter in self._adapter_cache.values():
            coros.append(adapter.close())
        for generic in self._generic_cache.values():
            coros.append(generic.close())
        if coros:
            await asyncio.gather(*coros, return_exceptions=True)
        self._adapter_cache.clear()
        self._generic_cache.clear()


class _PlatformAccess:
    """Convenience accessor: sdk.platforms.instagram etc."""

    def __init__(self, sdk: SocialSDK):
        self._sdk = sdk

    @property
    def instagram(self) -> BasePlatform:
        return self._sdk.platform("instagram")

    @property
    def tiktok(self) -> BasePlatform:
        return self._sdk.platform("tiktok")

    @property
    def x(self) -> BasePlatform:
        return self._sdk.platform("x")

    @property
    def facebook(self) -> BasePlatform:
        return self._sdk.platform("facebook")

    @property
    def youtube(self) -> BasePlatform:
        return self._sdk.platform("youtube")

    @property
    def linkedin(self) -> BasePlatform:
        return self._sdk.platform("linkedin")

    @property
    def threads(self) -> BasePlatform:
        return self._sdk.platform("threads")
